File: leetcode/src/79_Word_Search.py
'''
79. Word Search

Given a 2D board and a word, find if the word exists in the grid.

The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.

Example:

board =
[
  ['A','B','C','E'],
  ['S','F','C','S'],
  ['A','D','E','E']
]

Given word = "ABCCED", return true.
Given word = "SEE", return true.
Given word = "ABCB", return false.

Constraints:

board and word consists only of lowercase and uppercase English letters.
1 <= board.length <= 200
1 <= board[i].length <= 200
1 <= word.length <= 10^3

'''

class Solution(object):
    def exist(self, board, word):
        for i, row in enumerate(board):
            for j, cur_char in enumerate(row):
                if cur_char == word[0]:
                    visited = set()
                    visited.add((i,j))
                    if self.traverse(i, j, board, word, 0, visited) == True:
                        return True
        return False

    def traverse(self, x, y, board, word, word_idx, visited):
        if board[x][y] != word[word_idx]:
            return False
        if board[x][y] == word[word_idx] and word_idx == len(word) - 1:
            return True
        for (dir_x, dir_y) in [(0,1), (0,-1), (1,0),(-1,0)]:
            cur_x, cur_y = dir_x + x, dir_y + y
            if 0 <= cur_x < len(board) and 0 <= cur_y < len(board[0]) and not (cur_x, cur_y) in visited:
                visited.add((cur_x, cur_y))
                if self.traverse(cur_x, cur_y, board, word, word_idx+1, visited) == True:
                    return True
                visited.remove((cur_x, cur_y))

        return False

File: leetcode/src/79_wordSearch.js
/*
Given a 2D board and a word, find if the word exists in the grid.

The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.

Example:

board =
[
  ['A','B','C','E'],
  ['S','F','C','S'],
  ['A','D','E','E']
]

Given word = "ABCCED", return true.
Given word = "SEE", return true.
Given word = "ABCB", return false.
*/

/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */

const exist = (board, word) => {

  // found value;
  let found = false;
  
  // loop through to find a value;
  for (let row = 0; row < board.length; row++){
    for (let col = 0; col < board[row].length; col++){
      if(board[row][col] == word[0]){
        dfs(row, col, 0);
      }
    }
  }

  // Create dfs function
  function dfs(row, col, count){
    // goal
    if(count == word.length){
      found = true;
      return;
    }
    
    // check edge cases
    if (
      row < 0 ||
      row >= board.length ||
      col < 0 ||
      col >= board[0].length ||
      board[row][col] !== word[count] ||
      found
      ) {
        return;
      }

    let temp = board[row][col];
    board[row][col] = null;

    dfs (row+1, col, count+1)
    dfs (row-1, col, count+1)
    dfs (row, col+1, count+1)
    dfs (row, col-1, count+1)

    board[row][col] = temp;
    return
  }

  return found;
}

/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */

var exist = function (board, word) {
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[0].length; j++) {
      if (DFS(board, i, j, word, 0)) return true;
    }
  }
  return false;
};

function DFS(board, i, j, word, idx) {
  if (idx === word.length - 1 && board[i][j] === word[idx]) return true;
  if (board[i][j] !== word[idx]) return false;

  const dirs = [[0, 1], [0, -1], [-1, 0], [1, 0]];
  let cur = board[i][j];
  board[i][j] = "*";
  for (let k = 0; k < dirs.length; k++) {
    const [dirX, dirY] = dirs[k];
    const [curX, curY] = [dirX + i, dirY + j];
    if (curX < 0 || curY < 0 || curX >= board.length || curY >= board[0].length || board[curX][curY] === "*") continue;
    if (DFS(board, curX, curY, word, idx + 1)) return true;
  }
  board[i][j] = cur;
  return false;
}